### PR TITLE
Wake Centerline SDF: wake-structure proximity features for tandem

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The model has wake deficit features (gap-normalized TE offsets, merged PR #2213), but NO explicit feature encoding a node's distance and angle relative to the fore-foil WAKE CENTERLINE — the ray extending downstream from the TE along the freestream direction. In tandem configs, the aft-foil operates inside this wake. The wake centerline is the locus of maximum velocity deficit and turbulence intensity. Each node's position relative to the wake structure fundamentally determines its pressure:
- Above centerline → reduced momentum flow → different pressure
- Below centerline → near-freestream flow → cleaner pressure
- Crossing the centerline → maximum disturbance zone

This is COMPLEMENTARY to the existing wake deficit features: wake deficit encodes position relative to the TE *point*; wake centerline SDF encodes position relative to the extended wake *structure*.

## Instructions

Add `--wake_centerline_feature` flag. 4 new input channels: (d_perp, d_parallel, d_to_edge, inside_wake). Zero for single-foil samples.

### Implementation

```python
def compute_wake_features(pos, fore_te, aoa_rad, chord=1.0):
    """Wake centerline proximity features for each mesh node.
    pos:      [B, N, 2] node positions
    fore_te:  [B, 2] fore-foil TE position (reuse TE detection from te_coord_frame)
    aoa_rad:  [B] AoA in radians
    Returns:  [B, N, 4] (d_perp, d_parallel, d_to_edge, inside_wake)
    """
    # Wake propagates in freestream direction from fore TE
    wake_dir = torch.stack([torch.cos(aoa_rad), torch.sin(aoa_rad)], dim=-1)  # [B, 2]
    
    # Vector from fore TE to each node
    r = pos - fore_te.unsqueeze(1)  # [B, N, 2]
    
    # Parallel distance (downstream of TE) — clamp at 0 (only downstream matters)
    d_parallel = (r * wake_dir.unsqueeze(1)).sum(dim=-1).clamp(min=0)  # [B, N]
    
    # Perpendicular (signed) distance to centerline
    perp_dir = torch.stack([-wake_dir[:, 1], wake_dir[:, 0]], dim=-1)  # [B, 2]
    d_perp = (r * perp_dir.unsqueeze(1)).sum(dim=-1)  # [B, N] signed
    
    # Wake width model: classic turbulent spreading ~0.1 * sqrt(x/c)
    wake_width = 0.1 * torch.sqrt(d_parallel / chord + 1e-3)  # [B, N]
    
    # Distance to wake edge (positive = outside, negative = inside)
    d_to_edge = (d_perp.abs() - wake_width) / chord  # [B, N]
    inside_wake = (d_perp.abs() < wake_width).float()  # [B, N]
    
    # Normalize and stack
    feats = torch.stack([
        d_perp / chord,       # signed perpendicular distance (normalized)
        d_parallel / chord,    # downstream distance (normalized)
        d_to_edge,            # distance to wake edge (normalized)
        inside_wake           # binary: inside wake zone
    ], dim=-1)  # [B, N, 4]
    
    return feats
```

### Integration

1. Reuse the TE detection logic from `--te_coord_frame` (PR #2207) to get `fore_te` coordinates.
2. Extract AoA from input features.
3. Compute wake features and concatenate to input:
```python
if args.wake_centerline_feature:
    wake_feats = compute_wake_features(pos, fore_te, aoa_rad)
    # Zero out for single-foil samples
    wake_feats = wake_feats * is_tandem.unsqueeze(-1).unsqueeze(-1)
    x = torch.cat([x, wake_feats], dim=-1)  # +4 channels
    input_dim += 4
```

### Flags
```python
parser.add_argument('--wake_centerline_feature', action='store_true',
                    help='Add 4-channel wake centerline SDF features for tandem configs')
```

### Training commands
```bash
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent askeladd --seed 42 \
  --wandb_name "askeladd/wake-centerline-s42" \
  --wandb_group "wake-centerline-sdf" \
  --wake_centerline_feature \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0
# Seed 73: same but CUDA_VISIBLE_DEVICES=1, --seed 73
```

## Baseline
| Metric | 2-seed avg | Target |
|--------|-----------|--------|
| **p_in** | **11.742** | < 11.74 |
| p_oodc | 7.643 | < 7.64 |
| **p_tan** | **27.874** | < 27.87 |
| p_re | 6.419 | < 6.42 |

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Wake Centerline SDF (4 extra input channels)

| Metric | Baseline | S42 | S73 | 2-seed Avg | Delta |
|--------|----------|-----|-----|------------|-------|
| **p_in** | 11.742 | 12.542 | 12.056 | 12.299 | **+4.7%** |
| **p_oodc** | 7.643 | 7.477 | 7.751 | 7.614 | **-0.4%** |
| **p_re** | 6.419 | 6.258 | 6.326 | 6.292 | **-2.0%** |
| **p_tan** | 27.874 | 29.014 | 28.724 | 28.869 | **+3.6%** |

**W&B Run IDs:**
- S42: `agl3vast` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/agl3vast
- S73: `whbwg013` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/whbwg013

**Command used:**
```bash
CUDA_VISIBLE_DEVICES=<0|1> python train.py --agent askeladd --seed <42|73> \
  --wandb_name "askeladd/wake-centerline-s<seed>" --wandb_group "wake-centerline-sdf" \
  --wake_centerline_feature \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0
```

**Peak memory:** ~55 GB (GPU 0), ~83 GB (GPU 1)
**Epochs:** 500 (full run)

### Analysis

**Overall: Mixed result, leaning negative.** The wake centerline SDF features hurt in-distribution (-4.7%) and tandem transfer (-3.6%), while showing modest improvement on OOD-Re (+2.0%) and being essentially flat on OOD-Cond (-0.4%).

**What happened:**
1. **In-dist regression (+4.7%):** The 4 extra input channels (d_perp, d_parallel, d_to_edge, inside_wake) add parameters to the first projection layer. For single-foil samples (~87% of training data), these channels are all zeros, which means the model has to learn to ignore them for most samples while using them for the tandem minority. This creates an optimization tension.

2. **Tandem regression (+3.6%):** Surprising — we expected tandem to benefit most since these features encode wake structure explicitly. Possible causes:
   - The hard binary `inside_wake` feature and fixed spreading coefficient (0.1) may be too rigid. The actual wake structure varies with Re and AoA, and a fixed turbulent spreading model may introduce more noise than signal.
   - The fore-foil TE detection (max x-coord among surface nodes with saf_norm <= 0.005) may not always identify the correct TE, especially for different foil geometries.
   - The model already has wake_deficit_feature which encodes similar (but different) wake proximity information — the two features may be redundant or conflicting.

3. **OOD-Re improvement (-2.0%):** The wake features provide a useful geometric prior for Reynolds number generalization, since wake structure scales with Re. This is the one split where explicit wake geometry helps.

**Why it likely didn't work:**
- The existing wake_deficit_feature already provides the model with wake proximity information. Adding a second wake parameterization with a different geometric model creates redundancy.
- The turbulent spreading model (width = 0.1 * sqrt(x/c)) is a crude approximation that doesn't account for Re-dependent wake evolution or pressure gradients.
- 4 extra channels on a 66→70 dim input is a ~6% increase in input dimensionality, all of which are zero for single-foil samples.

### Suggested follow-ups
1. **Softer wake boundary:** Replace the hard binary `inside_wake` with a smooth sigmoid transition, and increase the spreading coefficient to 0.15. This was suggested by the research sub-agent but couldn't be tested since experiments were already launched.
2. **Wake features only for tandem:** Apply the wake features only when the sample is tandem (they're already zeroed for single-foil, but the projection weights still see zeros). Could try masking the gradient for the wake channels on single-foil samples.
3. **Replace wake_deficit_feature:** Instead of stacking both wake features, try using wake_centerline as a *replacement* for wake_deficit_feature to reduce redundancy.
4. **Re-dependent wake width:** Make the spreading coefficient a function of Re instead of a fixed constant.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2303 because ADVISOR: Closing — wake centerline SDF results are net-negative (p_in +4.7%, p_tan +3.6%), despite the modest p_re improvement (-2.0%).

**Root cause:** The wake_deficit_feature (merged PR #2213) already captures the essential wake proximity information. Adding a second wake parameterization with a fixed turbulent spreading model (width = 0.1·√(x/c)) introduces more noise than signal:
1. The 4 zero-valued channels for single-foil samples (~87% of data) create optimization tension in the input projection
2. The hard binary inside_wake and fixed spreading coefficient are too rigid — actual wake structure varies with Re and AoA
3. Redundancy with wake_deficit_feature causes conflicting gradient signals

The p_re gain suggests that geometric wake priors DO help OOD-Re generalization, but not enough to justify the in-dist and tandem regressions. This insight will be fed into future directions.

Next assignment coming shortly..
